### PR TITLE
android_native_app_glue.c translation to D

### PR DESCRIPTION
**Pull request by buggins**
It can help to avoid .c compilation, and probably allows to link directly from ldc2 (all in once)

### CODE REVIEW

**Comment by joakim-noah** [comment]
Thanks, I don't think a full translation to D will work, because `rt_init` and `rt_term` need to be called before D code calls the GC.  However, I have not played with the boundary enough to know where it is.

`rt_init` and `rt_term` are not necessary for the given C sample apps that have been translated to D: leave them out of the C version of this wrapper and the app will still work.  However, once you start invoking the GC, say by running the phobos unit tests in the app, it will start crashing without those runtime functions.

Now, does that mean we can just add them to this D wrapper instead, before `android_main`?  I'm not sure, we'll have to dig more into how those runtime initialization and termination functions work and whether your D wrapper might do something that needs them to be called first, from some outside C wrapper.

If you want to dig into that, feel free, but it will need to be done before we can use this D wrapper.

**Comment by buggins** [comment]
I believe, if D code does not use GC nor thread local storage, it should be ok before rt_init

**Comment by joakim-noah** [comment]
OK, I just want to be sure that's not happening anywhere in your translated D wrapper.  Maybe try slapping all functions except `android_app_entry` with @nogc and see if it still works?

**Comment by buggins** [comment]
Works for me.
Although I did not mark rest of glue functions as system/nogc/safe.

I'm working on DlangUI support of Android.
Hope that it will be usable soon.

Are there any plans to include your Android patches for ldc, druntime and phobos upstreams?

**Comment by joakim-noah** [comment]
I'll try building and using this translated D wrapper this week and let you know about this PR.

What do you need for DlangUI support, everything working okay so far?  Anything you need JNI for, to call the Android APIs that are Java-only?  I'm looking at JNI next, getting some crashes so far.

As for pushing my patches upstream, most of it is now merged.  Take a look at my gists to see the much smaller remaining patches, which consist of details that are not worth merging.

**Comment by buggins** [comment]
For now, I will implement all possible functionality provided by NativeActivity.
Currently, DlangUI application window is visible after start but then clears screen in a few seconds.
Once this issue is resolved, I'll add touch events handling.

**Comment by buggins** [comment]
DlangUI is actually port of my cross-platform UI C++ library used for CoolReader GL application.
It was not using NativeActivity. Instead, it was passing all events from Java to JNI library.
If it's possible to access JNI from native code directly, it would be great.

**Comment by Laeeth** [comment]
https://github.com/Monnoroch/DJni
Not maintained but did you look at this ?

On 21 April 2016 08:18:43 BST, Vadim Lopatin <notifications@github.com> wrote:
>DlangUI is actually port of my cross-platform UI C++ library used for
>CoolReader GL application.
>It was not using NativeActivity. Instead, it was passing all events
>from Java to JNI library.
>If it's possible to access JNI from native code directly, it would be
>great.
>
>---
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/joakim-noah/android/pull/6#issuecomment-212781330

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by buggins** [comment]
Fixed weird bug with memset

**Comment by buggins** [comment]
Announce of Android support in DlangUI
http://forum.dlang.org/post/cdekkumjynhqoxvmgjze@forum.dlang.org

File: android_native_app_glue_impl.d [hunk 1]
```diff
@@ -261,8 +261,9 @@ extern(C) static void* android_app_entry(void* param) {
 
 static android_app* android_app_create(ANativeActivity* activity,
         void* savedState, size_t savedStateSize) {
-    android_app* android_app = cast(android_app*)malloc(android_app.sizeof);
-    memset(android_app, 0, android_app.sizeof);
+    size_t sz = android_app.sizeof;
+    android_app* android_app = cast(android_app*)malloc(sz);
+    memset(android_app, 0, sz);
```
Comment (joakim-noah) [inline android_native_app_glue_impl.d]: Why is this necessary, ie what goes wrong without it?  Seems like both should do the same thing.

**Comment by joakim-noah** [comment]
[I've translated the glue code from C to D myself](https://github.com/joakim-noah/android/commit/940bd899664db1f72a68d83bff197d55952089be).  It's pretty much the same as your PR, with some slight differences.  Let me know if you have any problems with it, particularly that memset workaround you used.  I suggest you try the ldc 1.1.0 beta when compiling.

**Comment by joakim-noah** [comment]
Let me know if my version causes any problems.
